### DISCUSSION TITLE
2021 Q3 Roadmap

### DISCUSSION THREAD

**Discussion by tobegit3hub** [4paradigm/OpenMLDB#201] [General]
# OneStore engine
* Import offline data
* Sync data between online and offline
* Manage and sync catalog 

# Feature improvement
* Bulk load
* Reslove memory fragement
* Generage DDL by SQL and schema
* Develop the script to build  demo automaticly

# Docs
* Deployment docs
* Usage docs
* Architecture and technical docs

# Batch Execution Engine
* Support automatically packing for OpenMLDB Spark distribution
* Support Spark distribution for Linux, MacOS and Allinone versions
* Support reading and write tables from catalog service like Hive metastore
* Integrated with data exploration service like Presto and Hive
* Resolve the issue to run WindowSkewOpt for multiple windows
* Make UnsafeRowOpt stable and use by default
* Fix all the static check warning and increase 30% unit tests code coverage 

# Online Execution Engine
* Support distributed execution with pulling data from multiple tablets.
* Support time column index optimization for filter and join operator
* Support parameterized query

# SQL ability
* Support more math, string, comparison, flow control and other functions.

# bugfix and some other things

**Comment by jingchen2222** [comment]
## OpenMLDB Execution Engine
https://github.com/4paradigm/OpenMLDB/projects/3

**Running Issues**
- 


### SQL ability
Support more math, string, comparison, flow control and other functions.
- (RUNNING) #398 
- #222 
- #365
- #228 #220
- #224 
- #227
- (RUNNING) #234 
- #235 

### Online Execution Engine

- [x] Support parameterized query: https://github.com/4paradigm/OpenMLDB/discussions/287
  - #262 
  - #167 
  - #290 

- [x] Support Distributed query on batch mode: https://github.com/4paradigm/OpenMLDB/discussions/320
 
  - (RUNNING) #219 
  -  #317 
  - #318 
  - #362 


- [ ] Support time column index optimization for filter and join operator
  - #221
- [ ] Resign the Error System #error
    - #229 
    - #406

### Benchmark
- [ ] #377 
  - https://github.com/4paradigm/OpenMLDB/issues/432


### SDK and Api Server
- #397 
- #426 

### build & workflow & style
- #239

**Comment by jingchen2222** [comment]
# SQL Ability

## Query

## Insert

## Create Table

## Create Procedure
